### ISSUE TITLE
Make parameters visible to input and output

### ISSUE THREAD

**Issue by snajder-r** [snakemake/snakemake#1122]
The **params** allow us define arbitrary parameters that can be referenced by **shell** and **run**.

The **input** and **output**, however, cannot refer to the parameters (and vice versa).

Sometimes be handy, though, if parameters could be used to construct strings (for example from wildcards) that are then repeatedly used in the input and/or output.

A very simple example would be the construction of an output directory, if i want to refer to both the output file and the directory. What I would like to do is:

```
from pathlib import Path

rule A:
    output: Path(params.outdir).joinpath("bar.txt")
    params:
        outdir="foo{some_wildcard}"
    shell: """
           some_setup_script {params.outdir} 
           something_that_produces {output} 
           some_cleanup_script {params.outdir} 
           """

rule test:
    output: expand(rules.A.output, some_wildcard="ooo")
```

Currently this produces the error:

```
NameError in line 4 of Snakefile:
name 'params' is not defined
  File "Snakefile", line 4, in <module>
```

The workaround I currently have to do is to copy the construction of the directory name (which, as all copy&paste jobs, is error prone):

```
from pathlib import Path

rule A:
    output: "foo{some_wildcard}/bar.txt"
    params:
        outdir="foo{some_wildcard}"
    shell: """
           some_setup_script {params.outdir}
           something_that_produces {output}
           some_cleanup_script {params.outdir}
           """

rule test:
    output: expand(rules.A.output, some_wildcard="ooo")
```

I realize this could raise the question why not to make it the other way around (let params refer to output) and we can't have both since we'd then have problems with circular definitions. I would argue, however, that some (documented) evaluation order (e.g. params -> input -> output -> shell) is better than no evaluation order. I also think that params are particularly well suited for constructing strings in advance and then letting the rest of the rule definition use them.

**Comment by aryarm** [comment]
Hi @snajder-r ,
You can refer to `output` in `params` using a lambda function. Here's a common refrain that I like to use in these situations.
```
params:
    outdir = lambda wildcards, output: Path(output.file).parent
```
Here's that snippet placed in the context of your example:
```
from pathlib import Path

rule A:
    output:
        bar = "foo{some_wildcard}/bar.txt"
    params:
        outdir = lambda wildcards, output: Path(output.bar).parent
    shell: """
           some_setup_script {params.outdir}
           something_that_produces {output.bar}
           some_cleanup_script {params.outdir}
           """

rule test:
    output: expand(rules.A.output, some_wildcard="ooo")
```

**Comment by snajder-r** [comment]
That's interesting. If I can refer to output in params using a lambda function, it should be doable to make it so that I can just refer to output straight away, right?

Personally, I would still prefer it the other way around (seems more logical to me to infer output from param rather than param from output), but I can see how that may be difficult if the opposite is already possible (and not creating circular references).

**Comment by aryarm** [comment]
Yeah, it would be nice if it went both ways. I'm no Snakemake expert, but I think the way that rule dependencies are resolved requires that the output be a string and not a lambda function. Like, from what I understand, when Snakemake is computing the DAG, it will first look at the output of each rule to see if it matches the requested output. And then it will fill the wildcards and resolve the input and params lambda functions of the rules that match up with those outputs. And then it will figure out which outputs match those resolved inputs, and so on.
So I think we can't provide a lambda function to the output because, like you mentioned, it's a chicken-and-egg problem. We can't resolve params without knowing the wildcards from matching the output.

**Comment by snajder-r** [comment]
Right, I see the point. This made me think of another possible workaround:

```
A_outdir = Path("foo{some_wildcard}")

rule A:
    output:
        bar = A_outdir.joinpath("bar.txt")
    params:
        outdir = A_outdir   # Here wildcards are replaced
    shell: """
           some_setup_script {params.outdir}
           something_that_produces {output.bar}
           some_cleanup_script {params.outdir}
           """
```

I don't like how it requires me to define something rule specific outside the rule, tough. It would be cool to be able to define some kind of constants within the rule, like this:


```
rule A:
   constants:
       outdir = Path("foo{some_wildcard}")
    output:
        bar = constants.outdir.joinpath("bar.txt")
    params:
        outdir = constants.A_outdir  # Here wildcards are replaced
    shell: """
           some_setup_script {params.outdir}
           something_that_produces {output.bar}
           some_cleanup_script {params.outdir}
           """
```

**Comment by aryarm** [comment]
> This made me think of another possible workaround

Yup, I think that should work, too!

**Comment by codykingham** [comment]
Ditto. Params in output please.

**Comment by RyanVidegar-Laird** [comment]
Accessing params in output would be very convenient. Please implement!

**Comment by aryarm** [comment]
Hopefully someone can figure out how to resolve the circularity issue and make params available in output

@tgelafr1, in the meantime, you might be able to achieve your behavior by accessing output in params

For example, perhaps something like this?
```diff
+ from pathlib import Path

rule process_data:
    params:
-         output_dir="somedir/post_processing"
+         output_dir=lambda w, output: Path(output[0]).parent
    output:
-        "{params.output_dir}/{sample}.csv"
+        "somedir/post_processing/{sample.csv}"
    shell:
        "somecommand -o {params.output_dir}"

use rule process_data as process_pre_data with:
-    params:
-        output_dir="somedir/pre_processing"
+    output:
+        "somedir/pre_processing/{sample}.csv
```

**Comment by pkerbs** [comment]
Is there some progress on this?
I agree that this would be useful!

**Comment by AishaShah** [comment]
Is there any progress on this?
